pkg/agent/client/client.go

```diff
@@ -0,0 +1,213 @@
+	return &client{
+		c: c,
+	}
+}
+
+func newGRPCConn(c *Config) (*grpc.ClientConn, error) {
+	credFunc := func() (credentials.TransportCredentials, error) {
+		var tlsCert []tls.Certificate
+		var tlsConfig *tls.Config
+
+		svid, key, bundle := c.KeysAndBundle()
+		spiffePeer := &spiffe_tls.TLSPeer{
+			SpiffeIDs:  []string{"spiffe://" + c.TrustDomain.Host + "/spiffe/server"},
+			TrustRoots: util.NewCertPool(bundle...),
+		}
+		tlsCert = append(tlsCert, tls.Certificate{Certificate: [][]byte{svid.Raw}, PrivateKey: key})
+		tlsConfig = spiffePeer.NewTLSConfig(tlsCert)
+		return credentials.NewTLS(tlsConfig), nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second) // TODO: Make this timeout configurable?
+	defer cancel()
+
+	config := grpcutil.GRPCDialerConfig{
+		Log:      grpcutil.LoggerFromFieldLogger(c.Log),
+		CredFunc: credFunc,
+	}
+	dialer := grpcutil.NewGRPCDialer(config)
+	conn, err := dialer.Dial(ctx, c.Addr)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create connection: %v", err)
+	}
+	return conn, nil
+}
+
+func (c *client) FetchUpdates(req *node.FetchSVIDRequest) (*Update, error) {
+	nodeClient, err := c.newNodeClient()
+	if err != nil {
+		return nil, err
+	}
+
+	stream, err := nodeClient.FetchSVID(context.Background())
+	// We weren't able to get a stream...close the client and return the error.
+	if err != nil {
+		c.Release()
+		c.c.Log.Errorf("%v: %v", ErrUnableToGetStream, err)
+		return nil, ErrUnableToGetStream
+	}
+
+	// Send the request to the server using the stream.
+	err = stream.Send(req)
+	// Close the stream whether there was an error or not
+	stream.CloseSend()
+	if err != nil {
+		// TODO: should we try to create a new stream?
+		return nil, err
+	}
+
+	regEntries := map[string]*common.RegistrationEntry{}
+	svids := map[string]*node.Svid{}
+	var lastBundle []byte
+	// Read all the server responses from the stream.
+	for {
+		resp, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			if len(regEntries) > 0 || len(svids) > 0 || lastBundle != nil {
+				// There was an error receiving a response, exit loop to return what we have.
+				return &Update{regEntries, svids, lastBundle}, ErrPartialResponse
+			}
+
+			return nil, err
+		}
+
+		for _, re := range resp.SvidUpdate.RegistrationEntries {
+			regEntries[re.EntryId] = re
+		}
+		for spiffeid, svid := range resp.SvidUpdate.Svids {
+			svids[spiffeid] = svid
+		}
+		lastBundle = resp.SvidUpdate.Bundle
+	}
+	return &Update{regEntries, svids, lastBundle}, nil
+}
+
+func (c *client) Release() {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.conn != nil {
+		c.conn.Close()
+		c.conn = nil
+	}
+}
+
+func (c *client) newNodeClient() (node.NodeClient, error) {
+	if c.newNodeClientCallback != nil {
+		return c.newNodeClientCallback()
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.conn == nil {
+		conn, err := newGRPCConn(c.c)
+		if err != nil {
+			return nil, err
+		}
+		c.conn = conn
```

Comment:
Is there a reason to store conn on the struct instead of the node client itself? Feels like it might be simpler to store the client directly - we could avoid the callback in that case as the test suite can set a mocked client onto the struct

Reply:
If I hadn't conn, I couldn't close it when Release() is called. AFAIK I cannot get then connection from the node.NodeClient interface.